File: src/Sample/Plumix.Sample/Demos/Material/SnackBarDemoPage.cs
using System;
using Avalonia.Media;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.Widgets;

namespace Plumix;

// Dart parity source: dart_sample/lib/demos/material/snack_bar_demo_page.dart

public sealed class SnackBarDemoPage : StatefulWidget
{
    public override State CreateState() => new SnackBarDemoPageState();

    private sealed class SnackBarDemoPageState : State
    {
        private bool _floating;
        private bool _showAction = true;
        private bool _showCloseIcon;
        private bool _customColors;
        private bool _swipeUp;
        private bool _fixedWidth;
        private int _actionCount;

        public override Widget Build(BuildContext context)
        {
            return new Column(
                crossAxisAlignment: CrossAxisAlignment.Stretch,
                spacing: 14,
                children:
                [
                    new Text("SnackBar + SnackBarAction", fontSize: 20),
                    new Text(
                        "ScaffoldMessenger queue, fixed/floating placement, one-shot action, close icon, overflow, and theme/widget precedence.",
                        fontSize: 14,
                        color: Colors.DimGray),
                    new Row(
                        spacing: 8,
                        children:
                        [
                            ControlButton(_floating ? "Floating" : "Fixed", () => SetState(() => _floating = !_floating)),
                            ControlButton(_showAction ? "Action on" : "Action off", () => SetState(() => _showAction = !_showAction)),
                            ControlButton(_showCloseIcon ? "Close on" : "Close off", () => SetState(() => _showCloseIcon = !_showCloseIcon)),
                        ]),
                    new Row(
                        spacing: 8,
                        children:
                        [
                            ControlButton(
                                _customColors ? "Colors on" : "Colors off",
                                () => SetState(() => _customColors = !_customColors)),
                            ControlButton(
                                _swipeUp ? "Swipe up" : "Swipe down",
                                () => SetState(() => _swipeUp = !_swipeUp)),
                            ControlButton(
                                _fixedWidth ? "Width 320" : "Full width",
                                () => SetState(() => _fixedWidth = !_fixedWidth)),
                        ]),
                    new Row(
                        spacing: 8,
                        children:
                        [
                            new ElevatedButton(new Text("SHOW SNACKBAR"), () => ShowSnackBar(context)),
                        ]),
                    new Text($"Action presses: {_actionCount}", fontSize: 13),
                    new Text(
                        "Narrow the window to move a wide action below the message; repeated SHOW calls "
                        + "show queue order. Drag the bar in the swipe direction to dismiss it; the width "
                        + "probe needs floating behavior.",
                        fontSize: 13,
                        color: Colors.DimGray),
                ]);
        }

        private void ShowSnackBar(BuildContext context)
        {
            ScaffoldMessenger.Of(context).ShowSnackBar(new SnackBar(
                content: new Text("A queued Material snackbar from Plumix."),
                behavior: _floating ? SnackBarBehavior.Floating : SnackBarBehavior.Fixed,
                backgroundColor: _customColors ? Color.Parse("#FF004D40") : null,
                closeIconColor: _customColors ? Color.Parse("#FFFFD54F") : null,
                showCloseIcon: _showCloseIcon,
                persist: _showAction || _showCloseIcon,
                dismissDirection: _swipeUp ? DismissDirection.Up : DismissDirection.Down,
                width: _floating && _fixedWidth ? 320 : null,
                action: _showAction
                    ? new SnackBarAction(
                        label: "UNDO LONG ACTION",
                        textColor: _customColors ? Color.Parse("#FFFFD54F") : null,
                        onPressed: () => SetState(() => _actionCount++))
                    : null));
        }

        private static Widget ControlButton(string label, Action onPressed) =>
            new TextButton(new Text(label, fontSize: 12), onPressed);
    }
}

File: dart_sample/lib/demos/material/snack_bar_demo_page.dart
import 'package:material_ui/material_ui.dart';

class SnackBarDemoPage extends StatefulWidget {
  const SnackBarDemoPage({super.key});

  @override
  State<SnackBarDemoPage> createState() => _SnackBarDemoPageState();
}

class _SnackBarDemoPageState extends State<SnackBarDemoPage> {
  bool _floating = false;
  bool _showAction = true;
  bool _showCloseIcon = false;
  bool _customColors = false;
  bool _swipeUp = false;
  bool _fixedWidth = false;
  int _actionCount = 0;

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 14,
      children: <Widget>[
        const Text('SnackBar + SnackBarAction', style: TextStyle(fontSize: 20)),
        const Text(
          'ScaffoldMessenger queue, fixed/floating placement, one-shot action, close icon, overflow, and theme/widget precedence.',
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        Wrap(
          spacing: 8,
          runSpacing: 8,
          children: <Widget>[
            TextButton(
              onPressed: () => setState(() => _floating = !_floating),
              child: Text(_floating ? 'Floating' : 'Fixed'),
            ),
            TextButton(
              onPressed: () => setState(() => _showAction = !_showAction),
              child: Text(_showAction ? 'Action on' : 'Action off'),
            ),
            TextButton(
              onPressed: () => setState(() => _showCloseIcon = !_showCloseIcon),
              child: Text(_showCloseIcon ? 'Close on' : 'Close off'),
            ),
          ],
        ),
        Wrap(
          spacing: 8,
          runSpacing: 8,
          children: <Widget>[
            TextButton(
              onPressed: () => setState(() => _customColors = !_customColors),
              child: Text(_customColors ? 'Colors on' : 'Colors off'),
            ),
            TextButton(
              onPressed: () => setState(() => _swipeUp = !_swipeUp),
              child: Text(_swipeUp ? 'Swipe up' : 'Swipe down'),
            ),
            TextButton(
              onPressed: () => setState(() => _fixedWidth = !_fixedWidth),
              child: Text(_fixedWidth ? 'Width 320' : 'Full width'),
            ),
          ],
        ),
        Wrap(
          spacing: 8,
          runSpacing: 8,
          children: <Widget>[
            ElevatedButton(
              onPressed: () => _showSnackBar(context),
              child: const Text('SHOW SNACKBAR'),
            ),
          ],
        ),
        Text(
          'Action presses: $_actionCount',
          style: const TextStyle(fontSize: 13),
        ),
        const Text(
          'Narrow the window to move a wide action below the message; repeated SHOW calls '
          'show queue order. Drag the bar in the swipe direction to dismiss it; the width '
          'probe needs floating behavior.',
          style: TextStyle(fontSize: 13, color: Colors.black54),
        ),
      ],
    );
  }

  void _showSnackBar(BuildContext context) {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(
        content: const Text('A queued Material snackbar from Plumix.'),
        behavior: _floating
            ? SnackBarBehavior.floating
            : SnackBarBehavior.fixed,
        backgroundColor: _customColors ? const Color(0xFF004D40) : null,
        closeIconColor: _customColors ? const Color(0xFFFFD54F) : null,
        showCloseIcon: _showCloseIcon,
        persist: _showAction || _showCloseIcon,
        dismissDirection: _swipeUp
            ? DismissDirection.up
            : DismissDirection.down,
        width: _floating && _fixedWidth ? 320 : null,
        action: _showAction
            ? SnackBarAction(
                label: 'UNDO LONG ACTION',
                textColor: _customColors ? const Color(0xFFFFD54F) : null,
                onPressed: () => setState(() => _actionCount++),
              )
            : null,
      ),
    );
  }
}
